src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -317,9 +362,32 @@ public void onMessage(ConsumerRecord<K, V> record, Acknowledgment acknowledgment
 			}
 		}
 
+		@Override
+		public <T, E extends Throwable> boolean open(RetryContext context, RetryCallback<T, E> callback) {
+			if (KafkaMessageDrivenChannelAdapter.this.recoveryCallback != null) {
+				attributesHolder.set(context);
+			}
+			return true;
+		}
+
+		@Override
+		public <T, E extends Throwable> void close(RetryContext context, RetryCallback<T, E> callback,
+				Throwable throwable) {
+			if (KafkaMessageDrivenChannelAdapter.this.recoveryCallback != null) {
+				attributesHolder.remove();
```

Comment:
We remove the `ThreadLocal` variable when the `retryTemplate` is present.
But I don't see how it is clear otherwise.